File: bot/wikipedia_service.py
"""
Wikipedia API service for fetching celebrity images and information
"""

import aiohttp
import asyncio
import logging
from urllib.parse import quote
import re

logger = logging.getLogger(__name__)

class WikipediaService:
    def __init__(self):
        self.base_url = "https://en.wikipedia.org/w/api.php"
        self.headers = {
            'User-Agent': 'ArabCelebrityQuizBot/1.0 (Discord Bot for Educational Purposes)'
        }
        self.session = None

    async def get_session(self):
        """Get or create aiohttp session"""
        if self.session is None or self.session.closed:
            self.session = aiohttp.ClientSession(headers=self.headers)
        return self.session

    async def close_session(self):
        """Close aiohttp session"""
        if self.session and not self.session.closed:
            await self.session.close()

    async def search_celebrities(self, query, limit=20):
        """Search for celebrities and get their images"""
        session = await self.get_session()
        
        params = {
            "action": "query",
            "generator": "search",
            "gsrnamespace": 0,
            "gsrlimit": limit,
            "prop": "pageimages|extracts",
            "pilimit": "max",
            "exintro": True,
            "exsentences": 1,
            "exlimit": "max",
            "pithumbsize": 300,
            "format": "json",
            "gsrsearch": query
        }
        
        try:
            async with session.get(self.base_url, params=params) as response:
                if response.status == 200:
                    data = await response.json()
                    return self.parse_search_results(data)
                else:
                    logger.error(f"Wikipedia search failed with status {response.status}")
                    return []
        except Exception as e:
            logger.error(f"Error searching Wikipedia: {e}")
            return []

    def parse_search_results(self, data):
        """Parse Wikipedia search results"""
        results = []
        
        if 'query' in data and 'pages' in data['query']:
            for page_id, page_data in data['query']['pages'].items():
                if page_id == '-1':  # Skip missing pages
                    continue
                
                result = {
                    'name': page_data.get('title', ''),
                    'description': page_data.get('extract', ''),
                    'image_url': None
                }
                
                # Get image URL
                if 'thumbnail' in page_data:
                    result['image_url'] = page_data['thumbnail']['source']
                elif 'original' in page_data:
                    result['image_url'] = page_data['original']['source']
                
                results.append(result)
        
        return results

    async def get_celebrity_image(self, celebrity_name):
        """Get main image for a specific celebrity"""
        session = await self.get_session()
        
        params = {
            "action": "query",
            "prop": "pageimages",
            "format": "json",
            "piprop": "original",
            "titles": celebrity_name,
            "pithumbsize": 500
        }
        
        try:
            async with session.get(self.base_url, params=params) as response:
                if response.status == 200:
                    data = await response.json()
                    
                    if 'query' in data and 'pages' in data['query']:
                        page_data = list(data['query']['pages'].values())[0]
                        
                        if 'original' in page_data:
                            return page_data['original']['source']
                        elif 'thumbnail' in page_data:
                            return page_data['thumbnail']['source']
                
                logger.warning(f"No image found for {celebrity_name}")
                return None
                
        except Exception as e:
            logger.error(f"Error getting celebrity image: {e}")
            return None

    async def get_celebrity_profile(self, name):
        """Get complete celebrity profile with image, bio, and links"""
        session = await self.get_session()
        
        params = {
            "action": "query",
            "prop": "info|extracts|pageimages",
            "inprop": "url",
            "exsentences": 3,
            "titles": name,
            "format": "json",
            "pithumbsize": 400,
            "piprop": "original"
        }
        
        try:
            async with session.get(self.base_url, params=params) as response:
                if response.status == 200:
                    data = await response.json()
                    return self.parse_celebrity_profile(data)
                else:
                    logger.error(f"Failed to get profile for {name}")
                    return None
                    
        except Exception as e:
            logger.error(f"Error getting celebrity profile: {e}")
            return None

    def parse_celebrity_profile(self, data):
        """Parse celebrity profile data from Wikipedia API response"""
        profile = {}
        
        try:
            if 'query' in data and 'pages' in data['query']:
                page_data = list(data['query']['pages'].values())[0]
                
                # Skip if page doesn't exist
                if 'missing' in page_data:
                    return None
                
                profile['name'] = page_data.get('title', '')
                profile['description'] = page_data.get('extract', '')
                profile['wikipedia_url'] = page_data.get('fullurl', '')
                
                # Get image URL
                if 'original' in page_data:
                    profile['image_url'] = page_data['original']['source']
                elif 'thumbnail' in page_data:
                    profile['image_url'] = page_data['thumbnail']['source']
                else:
                    profile['image_url'] = None
                
                # Clean description
                if profile['description']:
                    profile['description'] = self.clean_description(profile['description'])
                
        except Exception as e:
            logger.error(f"Error parsing celebrity profile: {e}")
            return None
        
        return profile

    def clean_description(self, description):
        """Clean Wikipedia description text"""
        # Remove HTML tags
        description = re.sub(r'<[^>]+>', '', description)
        
        # Remove citation markers
        description = re.sub(r'\[\d+\]', '', description)
        
        # Remove extra whitespace
        description = re.sub(r'\s+', ' ', description).strip()
        
        # Limit length
        if len(description) > 300:
            description = description[:300] + "..."
        
        return description

    async def search_celebrity_variations(self, celebrity_name, aliases=None):
        """Search for celebrity using multiple name variations"""
        search_terms = [celebrity_name]
        
        if aliases:
            search_terms.extend(aliases)
        
        # Add common variations for Arab names
        variations = self.generate_name_variations(celebrity_name)
        search_terms.extend(variations)
        
        best_result = None
        best_score = 0
        
        for term in search_terms:
            try:
                profile = await self.get_celebrity_profile(term)
                if profile and profile.get('image_url'):
                    # Simple scoring based on name similarity
                    score = self.calculate_name_similarity(celebrity_name, profile['name'])
                    
                    if score > best_score:
                        best_score = score
                        best_result = profile
                        
                # Add small delay to respect rate limits
                await asyncio.sleep(0.1)
                
            except Exception as e:
                logger.error(f"Error searching for variation '{term}': {e}")
                continue
        
        return best_result

    def generate_name_variations(self, name):
        """Generate common variations for Arab names"""
        variations = []
        
        # Common transliteration variations
        replacements = {
            'mohammed': ['muhammad', 'mohamed'],
            'muhammad': ['mohammed', 'mohamed'],
            'ahmed': ['ahmad'],
            'ahmad': ['ahmed'],
            'omar': ['umar'],
            'umar': ['omar'],
            'hassan': ['hasan'],
            'hasan': ['hassan'],
            'hussein': ['husayn', 'hussain'],
            'husayn': ['hussein', 'hussain']
        }
        
        name_lower = name.lower()
        for original, variants in replacements.items():
            if original in name_lower:
                for variant in variants:
                    new_name = name_lower.replace(original, variant)
                    variations.append(new_name.title())
        
        # Add with/without common prefixes
        prefixes = ['al-', 'el-', 'abd', 'abdul']
        words = name.split()
        
        for prefix in prefixes:
            # Try adding prefix
            if not any(word.lower().startswith(prefix) for word in words):
                variations.append(f"{prefix.capitalize()} {name}")
            
            # Try removing prefix
            new_words = []
            for word in words:
                if word.lower().startswith(prefix):
                    new_words.append(word[len(prefix):])
                else:
                    new_words.append(word)
            
            if new_words != words:
                variations.append(' '.join(new_words))
        
        return list(set(variations))  # Remove duplicates

    def calculate_name_similarity(self, name1, name2):
        """Simple name similarity calculation"""
        name1 = name1.lower().strip()
        name2 = name2.lower().strip()
        
        if name1 == name2:
            return 100
        
        # Count common words
        words1 = set(name1.split())
        words2 = set(name2.split())
        
        common_words = words1.intersection(words2)
        total_words = words1.union(words2)
        
        if not total_words:
            return 0
        
        return (len(common_words) / len(total_words)) * 100

    async def __aenter__(self):
        """Async context manager entry"""
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Async context manager exit"""
        await self.close_session()

File: services/wikipediaService.js
/**
 * Wikipedia API service for fetching celebrity images and information
 */

const axios = require('axios');
const logger = require('../utils/logger');
const config = require('../config');

class WikipediaService {
    constructor() {
        this.baseUrl = "https://en.wikipedia.org/w/api.php";
        this.headers = {
            'User-Agent': config.WIKIPEDIA_USER_AGENT
        };
        this.rateLimitDelay = config.WIKIPEDIA_RATE_LIMIT * 1000; // Convert to milliseconds
        this.lastRequestTime = 0;
    }

    async rateLimit() {
        const now = Date.now();
        const timeSinceLastRequest = now - this.lastRequestTime;
        
        if (timeSinceLastRequest < this.rateLimitDelay) {
            const delay = this.rateLimitDelay - timeSinceLastRequest;
            await new Promise(resolve => setTimeout(resolve, delay));
        }
        
        this.lastRequestTime = Date.now();
    }

    async searchCelebrities(query, limit = 20) {
        await this.rateLimit();
        
        const params = {
            action: "query",
            generator: "search",
            gsrnamespace: 0,
            gsrlimit: limit,
            prop: "pageimages|extracts",
            pilimit: "max",
            exintro: true,
            exsentences: 1,
            exlimit: "max",
            pithumbsize: 300,
            format: "json",
            gsrsearch: query
        };
        
        try {
            const response = await axios.get(this.baseUrl, { 
                params, 
                headers: this.headers,
                timeout: 10000 
            });
            
            if (response.status === 200) {
                return this.parseSearchResults(response.data);
            } else {
                logger.error(`Wikipedia search failed with status ${response.status}`);
                return [];
            }
        } catch (error) {
            logger.error(`Error searching Wikipedia: ${error.message}`);
            return [];
        }
    }

    parseSearchResults(data) {
        const results = [];
        
        if (data.query && data.query.pages) {
            for (const [pageId, pageData] of Object.entries(data.query.pages)) {
                if (pageId === '-1') continue; // Skip missing pages
                
                const result = {
                    name: pageData.title || '',
                    description: pageData.extract || '',
                    image_url: null
                };
                
                // Get image URL
                if (pageData.thumbnail) {
                    result.image_url = pageData.thumbnail.source;
                } else if (pageData.original) {
                    result.image_url = pageData.original.source;
                }
                
                results.push(result);
            }
        }
        
        return results;
    }

    async getCelebrityImage(celebrityName) {
        await this.rateLimit();
        
        const params = {
            action: "query",
            prop: "pageimages",
            format: "json",
            piprop: "original",
            titles: celebrityName,
            pithumbsize: 500
        };
        
        try {
            const response = await axios.get(this.baseUrl, { 
                params, 
                headers: this.headers,
                timeout: 10000 
            });
            
            if (response.status === 200) {
                return this.parseImageResult(response.data);
            } else {
                logger.error(`Wikipedia image fetch failed with status ${response.status}`);
                return null;
            }
        } catch (error) {
            logger.error(`Error fetching celebrity image: ${error.message}`);
            return null;
        }
    }

    parseImageResult(data) {
        if (data.query && data.query.pages) {
            const pages = data.query.pages;
            const pageId = Object.keys(pages)[0];
            const pageData = pages[pageId];
            
            if (pageData.original) {
                return pageData.original.source;
            } else if (pageData.thumbnail) {
                return pageData.thumbnail.source;
            }
        }
        
        return null;
    }

    async getCelebrityProfile(name) {
        await this.rateLimit();
        
        const params = {
            action: "query",
            prop: "extracts|pageimages",
            format: "json",
            exintro: true,
            exsentences: 3,
            piprop: "original",
            titles: name,
            pithumbsize: 500
        };
        
        try {
            const response = await axios.get(this.baseUrl, { 
                params, 
                headers: this.headers,
                timeout: 10000 
            });
            
            if (response.status === 200) {
                return this.parseCelebrityProfile(response.data);
            } else {
                logger.error(`Wikipedia profile fetch failed with status ${response.status}`);
                return null;
            }
        } catch (error) {
            logger.error(`Error fetching celebrity profile: ${error.message}`);
            return null;
        }
    }

    parseCelebrityProfile(data) {
        if (data.query && data.query.pages) {
            const pages = data.query.pages;
            const pageId = Object.keys(pages)[0];
            const pageData = pages[pageId];
            
            if (pageId === '-1') return null; // Page not found
            
            return {
                name: pageData.title,
                description: this.cleanDescription(pageData.extract),
                image_url: pageData.original ? pageData.original.source : null,
                wikipedia_url: `https://en.wikipedia.org/wiki/${encodeURIComponent(pageData.title)}`
            };
        }
        
        return null;
    }

    cleanDescription(description) {
        if (!description) return '';
        
        // Remove HTML tags
        description = description.replace(/<[^>]*>/g, '');
        
        // Remove extra whitespace
        description = description.replace(/\s+/g, ' ').trim();
        
        // Limit length
        if (description.length > 300) {
            description = description.substring(0, 297) + '...';
        }
        
        return description;
    }

    async searchCelebrityVariations(celebrityName, aliases = []) {
        const variations = this.generateNameVariations(celebrityName);
        const allNames = [celebrityName, ...aliases, ...variations];
        const results = [];
        
        for (const name of allNames) {
            const searchResults = await this.searchCelebrities(name, 5);
            results.push(...searchResults);
        }
        
        // Remove duplicates based on name
        const uniqueResults = results.filter((result, index, self) => 
            index === self.findIndex(r => r.name === result.name)
        );
        
        return uniqueResults;
    }

    generateNameVariations(name) {
        const variations = [];
        
        // Common Arabic name variations
        const commonVariations = {
            'mohammed': 'muhammad',
            'mohamed': 'muhammad',
            'mohammad': 'muhammad',
            'ahmed': 'ahmad',
            'omar': 'umar',
            'hassan': 'hasan',
            'hussein': 'husayn',
            'hussain': 'husayn',
            'abdallah': 'abdullah',
            'abd allah': 'abdullah',
            'abd al': 'abdul',
            'abdel': 'abdul'
        };
        
        // Generate variations
        for (const [variant, standard] of Object.entries(commonVariations)) {
            if (name.toLowerCase().includes(variant)) {
                variations.push(name.toLowerCase().replace(variant, standard));
            }
        }
        
        // Add name with "al-" prefix
        if (!name.toLowerCase().startsWith('al ')) {
            variations.push(`al ${name}`);
        }
        
        // Add name without "al-" prefix
        if (name.toLowerCase().startsWith('al ')) {
            variations.push(name.substring(3));
        }
        
        return variations;
    }

    calculateNameSimilarity(name1, name2) {
        const normalize = (name) => name.toLowerCase().replace(/[^a-z\s]/g, '').trim();
        const n1 = normalize(name1);
        const n2 = normalize(name2);
        
        if (n1 === n2) return 100;
        
        // Simple similarity calculation
        const words1 = n1.split(' ');
        const words2 = n2.split(' ');
        
        let matches = 0;
        for (const word1 of words1) {
            for (const word2 of words2) {
                if (word1 === word2 || word1.includes(word2) || word2.includes(word1)) {
                    matches++;
                }
            }
        }
        
        return Math.round((matches / Math.max(words1.length, words2.length)) * 100);
    }
}

module.exports = WikipediaService; 